callhome.go

```diff
@@ -0,0 +1,157 @@
+package netconf
+
+import (
+	"crypto/tls"
+	"fmt"
+	"github.com/nemith/netconf/transport"
+	ncssh "github.com/nemith/netconf/transport/ssh"
+	nctls "github.com/nemith/netconf/transport/tls"
+	"golang.org/x/crypto/ssh"
+	"net"
+)
+
+type TransportType uint
+
+const (
+	TransportTypeSSH TransportType = iota
+	TransportTypeTLS
+)
+
+/*
+CallHome implements netconf callhome procedure as specified in RFC 8071
+*/
+type CallHome struct {
+	listener  net.Listener
+	network   string
+	addr      string
+	transport TransportType
+	configSSH *ssh.ClientConfig
+	configTLS *tls.Config
+	session   *Session
+}
+
+type CallHomeOption func(*CallHome)
+
+// WithAddress sets the address (as required by net.Listen) the CallHome server listen to
+func WithAddress(addr string) CallHomeOption {
+	return func(ch *CallHome) {
+		ch.addr = addr
+	}
+}
+
+// WithNetwork set the network (as required by net.Listen) the CallHome server listen to
+func WithNetwork(network string) CallHomeOption {
+	return func(ch *CallHome) {
+		ch.network = network
+	}
+}
+
+// WithTransport sets the transport type used by the CallHome server
+func WithTransport(t TransportType) CallHomeOption {
+	return func(ch *CallHome) {
+		ch.transport = t
+	}
+}
+
+// WithConfigSSH sets the ssh.ClientConfig used by the CallHome server to upgrade incoming connections
+func WithConfigSSH(sc *ssh.ClientConfig) CallHomeOption {
+	return func(ch *CallHome) {
+		ch.configSSH = sc
+	}
+}
+
+// WithConfigTLS sets the tls.Config used by the CallHome server to upgrade incoming connections
+func WithConfigTLS(tc *tls.Config) CallHomeOption {
+	return func(ch *CallHome) {
+		ch.configTLS = tc
+	}
+}
+
+// NewCallHome creates a CallHome client
+func NewCallHome(opts ...CallHomeOption) (*CallHome, error) {
+	const (
+		defaultAddress   = "0.0.0.0:4334"
+		defaultNetwork   = "tcp"
+		defaultTransport = TransportTypeSSH
+	)
+
+	ch := &CallHome{
+		addr:      defaultAddress,
+		network:   defaultNetwork,
+		transport: defaultTransport,
+	}
+
+	for _, opt := range opts {
+		opt(ch)
+	}
+
+	if ch.configTLS == nil && ch.configSSH == nil {
```

Comment:
In general if an option is mandatory then (by definition) it's not an option.  In general i try to avoid runtime errors like this based on functional options.

In my opinion It's better to provide multiple functions to construct the object with positional arguments than to do runtime checks like this